src/lib/Controller/util/g_image.c

```diff
@@ -298,129 +135,101 @@ static int g_image_png_save(GImage *img, const char *filename) {
-    free(data);
-  if (to_file)
-    fclose(f);
+    number_of_components = STBI_rgb;
+  if (stbi_write_png(filename, img->width, img->height, number_of_components, img->data, img->width * number_of_components) !=
+      1)
+    return -1;
   return 0;
 }
 
-static int g_image_tiff_save(GImage *img, const char *filename) {
-  TIFF *image;
-
-  image = TIFFOpen(filename, "wb");
-
-  if (image == NULL)
-    fprintf(stderr, "Unable to write TIFF file: %s\n", filename);
-
-  TIFFSetField(image, TIFFTAG_SAMPLEFORMAT, SAMPLEFORMAT_IEEEFP);
-  TIFFSetField(image, TIFFTAG_SMINSAMPLEVALUE, 0);
-  TIFFSetField(image, TIFFTAG_SMAXSAMPLEVALUE, 1);
-  TIFFSetField(image, TIFFTAG_IMAGEWIDTH, img->width);
-  TIFFSetField(image, TIFFTAG_IMAGELENGTH, img->height);
-  TIFFSetField(image, TIFFTAG_SAMPLESPERPIXEL, 1);
-  TIFFSetField(image, TIFFTAG_BITSPERSAMPLE, 32);
-  TIFFSetField(image, TIFFTAG_ROWSPERSTRIP, img->height);
-  TIFFSetField(image, TIFFTAG_ORIENTATION, (int)ORIENTATION_TOPLEFT);
-  TIFFSetField(image, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);
-  TIFFSetField(image, TIFFTAG_COMPRESSION, COMPRESSION_NONE);
-  TIFFSetField(image, TIFFTAG_PHOTOMETRIC, 1);
-
-  TIFFWriteEncodedStrip(image, 0, img->float_data, img->width * img->height * sizeof(float));
+static int g_image_jpeg_save(GImage *img, char quality, const char *filename) {
+  FILE *fd = fopen(filename, "wb");
+  if (!fd) {
+    fprintf(stderr, "Error: could not open \"%s\" for writing\n", filename);
+    return -1;
+  }
+  fclose(fd);
 
-  TIFFWriteDirectory(image);
+  if (stbi_write_jpg(filename, img->width, img->height, STBI_rgb, img->data, quality) != 1)
+    return -1;
+  return 0;
+}
 
-  TIFFClose(image);
+static int g_image_hdr_save(GImage *img, const char *filename) {
+  FILE *fd = fopen(filename, "wb");
+  if (!fd) {
+    fprintf(stderr, "Error: could not open \"%s\" for writing\n", filename);
+    return -1;
+  }
+  fclose(fd);
 
+  if (stbi_write_hdr(filename, img->width, img->height, STBI_grey, img->float_data) != 1)
+    return -1;
   return 0;
 }
 
 int g_image_save(GImage *img, const char *filename, char quality) {
   switch (g_image_get_type(filename)) {
     case G_IMAGE_JPEG:
-      return g_image_jpeg_save(img, quality, true, filename, NULL, NULL);
+      return g_image_jpeg_save(img, quality, filename);
     case G_IMAGE_PNG:
       return g_image_png_save(img, filename);
-    case G_IMAGE_TIFF:
-      return g_image_tiff_save(img, filename);
+    case G_IMAGE_HDR:
+      return g_image_hdr_save(img, filename);
     default:
       fprintf(stderr, "Cannot save: unsupported image type: %s\n", filename);
       return -1;
   }
 }
 
+struct ImageData {
+  unsigned char **target_data;
+  unsigned long *target_data_size;
+};
+
+void g_image_save_to_jpeg_buffer_callback(void *context, void *data, int size) {
+  if (!*(((struct ImageData *)context)->target_data))
+    *(((struct ImageData *)context)->target_data) = (unsigned char *)malloc(size);
```

Comment:
When I see a malloc and no corresponding free, I fear a memory leak... Are you sure this memory is free? Where? Can you add a comment in the code about it?

Reply:
Yes I am sure that the memory is free. As the user of this function has to provide the pointer he is then responsible to free it (this is not new and was already working the exact same way before). This is done here: https://github.com/cyberbotics/webots/blob/master/src/lib/Controller/api/default_robot_window.c#L507
But I will add a note explaining this in the header file next to the function.